### PR TITLE
chore(span_sampling): add jsonschema for span sampling

### PR DESCRIPTION
**Pull request by ZStriker19**
## Description
It was proposed by JD that I should add a schema rather than do a lot of validation for the span sampling json we pull via envar or file. 

## Checklist
- [ ] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] Ensure tests are passing for affected code.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: ddtrace/internal/sampling.py [hunk 1]
```diff
@@ -208,44 +211,102 @@ def apply_span_sampling_tags(self, span):
...
+        except Exception as e:
+            raise ValueError("Error creating single span sampling rule {}: {}".format(json.dumps(rule), e))
+        sampling_rules.append(sampling_rule)
+    return sampling_rules
+
+
+def validate_json(json_rules):
+    schema = {
+        "type": "array",
+        "anyOf": [
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:242]: I don't think you need this `anyOf`, since there is only one valid value... an array of objects.


```python
   schema = {
        "type": "array",
        "items": {
            "type": "object",
            "anyOf": [
                {"properties": {"service": {"type": "string"}}, "required": ["service"]},
                {"properties": {"name": {"type": "string"}}, "required": ["name"]},
            ],
            "properties": {"max_per_second": {"type": "integer"}, "sample_rate": {"type": "number"}},
        },
    }
```

Comment (ZStriker19) [reply ddtrace/internal/sampling.py:242]: Ah you're right, I thought that too, but couldn't figure out what the syntax would look like without it. Thanks!

File: ddtrace/internal/sampling.py [hunk 2]
```diff
@@ -208,44 +211,102 @@ def apply_span_sampling_tags(self, span):
...
+                sample_rate=sample_rate, service=service, name=name, max_per_second=max_per_second
+            )
+        except Exception as e:
+            raise ValueError("Error creating single span sampling rule {}: {}".format(json.dumps(rule), e))
+        sampling_rules.append(sampling_rule)
+    return sampling_rules
+
+
+def validate_json(json_rules):
+    schema = {
```
Comment (mabdinur) [inline ddtrace/internal/sampling.py:240]: Schema can be a constant

Comment (mabdinur) [reply ddtrace/internal/sampling.py:240]: We can pull it out and get rid of this function. Aside from that LGTM

File: ddtrace/internal/sampling.py [hunk 3]
```diff
@@ -73,6 +73,18 @@ class SamplingMechanism(object):
     total=False,
 )
 
+span_sampling_json_schema = {
```
Comment (mabdinur) [inline ddtrace/internal/sampling.py:76]:
```suggestion
SPAN_SAMPLING_JSON_SCHEMA = {
```

**Comment by mabdinur** [comment]
With the addition of the new package this test is failing: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/20247/workflows/9b33a31b-e2e2-4c9b-9e27-ddeec7800120/jobs/1371051. We'll need to update this case to handle/reformat `pkgutil_resolve_name`

File: tests/internal/test_packages.py [hunk 4]
```diff
@@ -17,8 +17,11 @@ def test_get_distributions():
         # The package name in typing_extensions-4.x.x.dist-info/METADATA is set to `typing_extensions`
         # this is inconsistent with the package name found in pkg_resources. The block below corrects this.
         # The correct package name is typing-extensions.
+        # The issue exists in pkgutil-resolve-name package.
         if pkg.name == "typing_extensions" and "typing-extensions" in pkg_resources_ws:
             importlib_pkgs.add("typing-extensions")
+        elif pkg.name == "pkgutil_resolve_name" and "pkgutil-resolve-name" in pkg_resources_ws:
+            importlib_pkgs.add("pkgutil-resolve-name")
```
Comment (brettlangdon) [inline tests/internal/test_packages.py:24]: this seems like an unrelated change?

or is this because we are adding `jsonschema`?

File: ddtrace/internal/sampling.py [hunk 5]
```diff
@@ -206,30 +218,22 @@ def apply_span_sampling_tags(self, span):
...
         # If max_per_second not specified default to no limit
-        max_per_second = int(rule.get("max_per_second", _SINGLE_SPAN_SAMPLING_MAX_PER_SEC_NO_LIMIT))
-        if service is not None and not isinstance(service, string_types):
-            raise ValueError("The service value is not a string or None:%r" % service)
-        if name is not None and not isinstance(name, string_types):
-            raise ValueError("The name value is not a string or None:%r" % name)
-
-        if service is None and name is None:
-            raise ValueError("Neither service or name specified for single span sampling rule:%r" % rule)
+        max_per_second = rule.get("max_per_second", -1)
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:236]:
```suggestion
        max_per_second = rule.get("max_per_second", _SINGLE_SPAN_SAMPLING_MAX_PER_SEC_NO_LIMIT)
```

?

File: tests/tracer/test_single_span_sampling_rules.py [hunk 6]
```diff
@@ -64,7 +65,7 @@ def test_sampling_rule_init_via_env():
 
     # Testing error thrown when neither name nor service is set
     with override_env(dict(DD_SPAN_SAMPLING_RULES='[{"sample_rate":1.0}]')):
-        with pytest.raises(ValueError):
+        with pytest.raises(ValidationError):
```
Comment (brettlangdon) [inline tests/tracer/test_single_span_sampling_rules.py:68]: are these errors descriptive/helpful?

do we want to catch and re-raise with a better error message?

File: tests/internal/test_packages.py (hunk 4 above)
Comment (ZStriker19) [reply tests/internal/test_packages.py:24]: It's because we're adding jsonschema, although somehow a bunch of tests are failing now?  https://github.com/DataDog/dd-trace-py/pull/4064#issuecomment-1211299845

File: tests/tracer/test_single_span_sampling_rules.py (hunk 6 above)
Comment (ZStriker19) [reply tests/tracer/test_single_span_sampling_rules.py:68]: Hmm only if you can read jsonschemas. I'll re-raise

File: ddtrace/internal/sampling.py [hunk 7]
```diff
@@ -206,30 +218,22 @@ def apply_span_sampling_tags(self, span):
...
         # Only set this tag if it's not the default -1
-        if self._max_per_second != -1:
+        if self._max_per_second != _SINGLE_SPAN_SAMPLING_MAX_PER_SEC_NO_LIMIT:
             span.set_metric(_SINGLE_SPAN_SAMPLING_MAX_PER_SEC, self._max_per_second)
 
 
 def get_span_sampling_rules():
     # type: () -> List[SpanSamplingRule]
     json_rules = _get_span_sampling_json()
+    validate(json_rules, SPAN_SAMPLING_JSON_SCHEMA)
```
Comment (P403n1x87) [inline ddtrace/internal/sampling.py:228]: do we have an idea of the performance impact of this validation?

**Review by P403n1x87 — APPROVED** [review]

File: ddtrace/internal/sampling.py (hunk 7 above)
Comment (ZStriker19) [reply ddtrace/internal/sampling.py:228]: Not currently, I can do that as follow up work as discussed internally.

**Review by Yun-Kim — APPROVED** [review]
